src/main.cpp

```diff
@@ -1988,40 +1985,43 @@ bool CheckProofOfResearch(
         return true;
 
     // 6-4-2017 - Verify researchers stored block magnitude
-    double dNeuralNetworkMagnitude = CalculatedMagnitude2(bb.cpid, block.nTime, false);
-    if( bb.Magnitude > 0
-        && (fTestNet || (!fTestNet && pindexPrev->nHeight > 947000))
-        && bb.Magnitude > (dNeuralNetworkMagnitude*1.25) )
+    double dNeuralNetworkMagnitude = CalculatedMagnitude2(cpid, block.nTime);
+
+    if (claim.m_magnitude > (dNeuralNetworkMagnitude*1.25)
+        && (fTestNet || (!fTestNet && pindexPrev->nHeight > 947000)))
     {
-        return error("CheckProofOfResearch: Researchers block magnitude > neural network magnitude: Block Magnitude %f, Neural Network Magnitude %f, CPID %s ",
-                     bb.Magnitude, dNeuralNetworkMagnitude, bb.cpid.c_str());
+        return error(
+            "CheckProofOfResearch: Researchers block magnitude > neural network magnitude: Block Magnitude %f, Neural Network Magnitude %f, CPID %s ",
+             claim.m_magnitude,
+             dNeuralNetworkMagnitude,
+             cpid);
     }
 
-    int64_t nCalculatedResearch = GetProofOfStakeReward(nCoinAge, nFees, bb.cpid, true, 1, block.nTime,
-                                                        pindexBest, "checkblock_researcher", OUT_POR, OUT_INTEREST, dAccrualAge, dMagnitudeUnit, dAvgMagnitude);
+    int64_t nCalculatedResearch = GetProofOfStakeReward(nCoinAge, nFees, cpid, true, 1, block.nTime,
+                                                        pindexBest, OUT_POR, OUT_INTEREST, dAccrualAge, dMagnitudeUnit, dAvgMagnitude);
 
     if(!IsV9Enabled_Tally(pindexPrev->nHeight))
     {
-        if (bb.ResearchSubsidy > ((OUT_POR*1.25)+1))
+        if (claim.m_research_subsidy > ((OUT_POR*1.25)+1))
         {
             if (fDebug) LogPrintf("CheckProofOfResearch: Researchers Reward Pays too much : Retallying : "
                                 "claimedand %f vs calculated StakeReward %f for CPID %s",
-                                bb.ResearchSubsidy, OUT_POR, bb.cpid);
+                                claim.m_research_subsidy, OUT_POR, cpid);
 
             TallyResearchAverages(pindexBest);
-            GetLifetimeCPID(bb.cpid);
-            nCalculatedResearch = GetProofOfStakeReward(nCoinAge, nFees, bb.cpid, true, 2, block.nTime,
-                                                        pindexBest, "checkblock_researcher_doublecheck", OUT_POR, OUT_INTEREST, dAccrualAge, dMagnitudeUnit, dAvgMagnitude);
+            GetLifetimeCPID(cpid);
+            nCalculatedResearch = GetProofOfStakeReward(nCoinAge, nFees, cpid, true, 2, block.nTime,
+                                                        pindexBest, OUT_POR, OUT_INTEREST, dAccrualAge, dMagnitudeUnit, dAvgMagnitude);
         }
     }
     (void)nCalculatedResearch;
 
-    if (bb.ResearchSubsidy > ((OUT_POR*1.25)+1))
+    if (claim.m_research_subsidy > ((OUT_POR*1.25)+1))
```

Comment:
And here. Perhaps we should create a ResearchValidate() which encapsulates the rules we want to apply wrt the tolerance allowed so we only have to keep track of it in one place. This code is desperately in need of a rules engine actually.

Maybe making these changes is best put in another PR, because this one is getting big enough, but we need to bite the bullet on this for Fern, which is going to be the only mandatory for a while.

Reply:
See #370.

Reply:
Agreed--I didn't want to change those rules in this PR. I tried to limit the scope to just the `hashBoinc` format changes this time around to keep the PR manageable. We can examine these rules in a follow-up.

Reply:
Agreed.